Firestore/core/src/firebase/firestore/nanopb/message.h

```diff
@@ -130,48 +155,63 @@ class Message {
   }
 
   /**
-   * Serializes this message into a byte buffer.
+   * Serializes this `Message` into a byte buffer.
    *
-   * The lifetime of the return value is entirely independent of this message.
+   * The lifetime of the return value is entirely independent of this `Message`.
    */
   grpc::ByteBuffer ToByteBuffer() const;
 
- private:
-  // For access to the explicit constructor. Most code shouldn't be able to
-  // modify the underlying proto.
-  friend class remote::WatchStreamSerializer;
-  friend class remote::WriteStreamSerializer;
-  friend class remote::DatastoreSerializer;
+  /**
+   * Serializes this `Message` into a `ByteString`.
+   *
+   * The lifetime of the return value is entirely independent of this `Message`.
+   */
+  ByteString ToByteString() const;
 
-  explicit Message(const pb_field_t* fields) : fields_{fields} {
+ private:
+  // Returns a pointer to the Nanopb-generated array that describes the fields
+  // of the Nanopb proto; the array is required to call most Nanopb functions.
+  //
+  // Note that this is essentially a property of the type, but cannot be made
+  // a template parameter for various technical reasons.
+  static const pb_field_t* fields() {
```

Comment:
Working on this PR, I came to the conclusion that the fields descriptor is a property of the type, not an object, so it makes more sense to not store it. It also makes `Message` safer because accidentally providing the wrong descriptor could lead to catastrophic results -- now, all the mappings between Nanopb types and corresponding descriptors are in one place that is much easier to check for correctness.

Ideally, this would be `constexpr`, but making Nanopb generate `constexpr` arrays under the C++11 constraints seems hard.